rootfs/etc/nginx/lua/balancer/leastconn.lua

```diff
@@ -0,0 +1,115 @@
+local util = require("util")
+local split = require("util.split")
+require("resty.core")
+
+local ngx = ngx
+local ipairs = ipairs
+local tostring = tostring
+local string = string
+local tonumber = tonumber
+local setmetatable = setmetatable
+local string_format = string.format
+local ngx_log = ngx.log
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+
+
+local _M = { name = "leastconn" }
+
+function _M.new(self, backend)
+    local o = {
+        peers = backend.endpoints
+    }
+    setmetatable(o, self)
+    self.__index = self
+    return o
+end
+
+function _M.is_affinitized()
+    return false
+end
+
+
+local function get_upstream_name(upstream)
+    return upstream.address .. ":" .. upstream.port
+end
+
+
+function _M.balance(self)
+    local peers = self.peers
+    local endpoint = peers[1]
+    local endpoints = ngx.shared.balancer_leastconn
+    local feasible_endpoints = {}
+
+    if #peers ~= 1 then
+        local lowestconns = 9999
+        -- find the lowest connection count
+        for _, peer in pairs(peers) do
+            local conns = endpoints:get(get_upstream_name(peer))
+            if conns == nil then
+              endpoints:set(get_upstream_name(peer),0,600)
```

Comment:
My concern with depending on the sync method to handle garbage collection is trusting that the sync method will be reliably called with updates, and that we won't lose the list of peers between updates. With that said the main reason I was thinking to distrust sync was in case of nginx restarts - but then in that case the shared dict would be reset anyway.
I think we should set the expiration time to infinite.

Reply:
I also agree the expiration time should be infinite. Do you have a reason to be suspicious of the sync method not being called reliably? I hope it is something we can depend on.

Reply:
No specific reason to be suspicious, call it a generalised sense of professional paranoia :). Will fix the expiration time shortly.